### ISSUE TITLE
Allow passing an alignment solution to tweakreg

### ISSUE THREAD

**Issue by stscijgbot-jp** [spacetelescope/jwst#7261]
_Issue [JP-2813](https://jira.stsci.edu/browse/JP-2813) was created on JIRA by [Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva):_

Allow an initial solution as input to tweakreg (shift, rotation, scale), with the option of not having tweakreg do any additional fitting but simply apply this solution directly to the input images.

**Comment by stscijgbot-jp** [comment]
_Comment by [Anton Koekemoer](https://jira.stsci.edu/secure/ViewProfile.jspa?name=koekemoe) on [JIRA](https://jira.stsci.edu/browse/JP-2813?focusedCommentId=605761#comment-605761):_

Let me suggest a solution for this particular ticket that does not involve any modifications to tweakreg, but instead could  be solved by a completely separate and simpler script.

This particular use case basically requests the ability to modify the WCS of a single calibrated image (eg a "_cal.fits" file) by supplying the following parameters:
 * delta_RA, delta_Dec, delta_PA_V3 (all specified in degrees, and essentially considered as incremental changes to the current RA, Dec and V3 angle of the reference pixel in the image)
 * delta_scale (specified as a multiplicative factor, where a value of 1.00 would produce no change at all)

This might be achieved by a fairly simple script – the main point would be that this script should transform and update the WCS as tweakreg does when it has found a solution, and it should make sure that the resulting output file has the updated WCS correctly saved in the model and in the header (this is the part that may be complicated for users to attempt, now that we have gone to a datamodel approach for storing WCS).

So if a simple script can be produced which will achieve the above, for a single calibrated image (eg *_cal.fits) then that may be a preferred solution for this particular ticket, and we could separate this ticket from its "parent" ticket JP-2620, and no changes in tweakreg might be needed for this.

**Comment by nden** [comment]
Fixed in #7430

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-2813?focusedCommentId=619748#comment-619748):_

Just to get the discussion on this going again, are folks OK with the approach suggested by [Anton Koekemoer](https://jira.stsci.edu/secure/ViewProfile.jspa?name=koekemoe) of creating a stand-alone script to do this, rather than adding yet more options into the `tweakreg` step? Should this be discussed by either the Cal WG or the DMS WG?  [Alicia Canipe](https://jira.stsci.edu/secure/ViewProfile.jspa?name=acanipe)

**Comment by stscijgbot-jp** [comment]
_Comment by [Mihai Cara](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mcara) on [JIRA](https://jira.stsci.edu/browse/JP-2813?focusedCommentId=619828#comment-619828):_

I am OK with the idea of the script. Not sure about "This might be achieved by a fairly simple script..." :) The script may end up being simple, but some computations will need to be made on how to convert input parameters to what is needed.

**Comment by stscijgbot-jp** [comment]
_Comment by [Alicia Canipe](https://jira.stsci.edu/secure/ViewProfile.jspa?name=acanipe) on [JIRA](https://jira.stsci.edu/browse/JP-2813?focusedCommentId=620036#comment-620036):_

Tagging [Swara Ravindranath](https://jira.stsci.edu/secure/ViewProfile.jspa?name=swara) as well, since Anton will be out for a while. 

By standalone script, does that mean something that would go to the data analysis tools? Should we ask [Duy Nguyen](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dnguyen)?

**Comment by stscijgbot-jp** [comment]
_Comment by [Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva) on [JIRA](https://jira.stsci.edu/browse/JP-2813?focusedCommentId=620076#comment-620076):_

[Alicia Canipe](https://jira.stsci.edu/secure/ViewProfile.jspa?name=acanipe) I think it can go in the jwst bin directory and be called directly, as a script. I am not aware of a package that holds jwst specific data analysis tools.

**Comment by stscijgbot-jp** [comment]
_Comment by [Alicia Canipe](https://jira.stsci.edu/secure/ViewProfile.jspa?name=acanipe) on [JIRA](https://jira.stsci.edu/browse/JP-2813?focusedCommentId=620103#comment-620103):_

Ah okay, thanks [Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva). Sorry, I wasn't sure what the options were for standalone scripts.

**Comment by stscijgbot-jp** [comment]
_Comment by [Anton Koekemoer](https://jira.stsci.edu/secure/ViewProfile.jspa?name=koekemoe) on [JIRA](https://jira.stsci.edu/browse/JP-2813?focusedCommentId=624100#comment-624100):_

thanks @⁣all for the comments, and I've also added a couple of extra details to the description just to help clarify it a bit more as well.

**Comment by stscijgbot-jp** [comment]
_Comment by [Anton Koekemoer](https://jira.stsci.edu/secure/ViewProfile.jspa?name=koekemoe) on [JIRA](https://jira.stsci.edu/browse/JP-2813?focusedCommentId=624469#comment-624469):_

Also following up on a discussion about how to save the outputs:
 * from the user's perspective it would be good if this script could write out a new image FITS file that's the same structure as the original one (eg, a new cal.fits file with all the original extensions present), but with the wcs model now updated. An argument called 'overwrite' (set to True/False) could control the behaviour of this (ie if set to False, then the user would need to specify an output filename as a parameter to the script).
 * in addition, if the user only wants to write out an asdf file with the new wcs model, then this could be provided as an extra option as well.

**Comment by stscijgbot-jp** [comment]
_Comment by [Mihai Cara](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mcara) on [JIRA](https://jira.stsci.edu/browse/JP-2813?focusedCommentId=624497#comment-624497):_

Do you envision this script to apply same deltas to multiple input files, that is should the script accept multiple input files? If so, specifying (multiple) output file names is quite difficult, i.e., the order is important and it easy to make mistakes and it is impossible to support (expand) wild characters in input file names. In that case it may be more helpful to provide suffix to be used for output file names (i.e., _cal -> _wcsadj or _tweakedwcs or something like that).

How would you prefer this script to work? Single files, multiple files? Output file names or suffixes?

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-2813?focusedCommentId=628130#comment-628130):_

Fixed by #7430

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-2813?focusedCommentId=629876#comment-629876):_

Reassigning to [Anton Koekemoer](https://jira.stsci.edu/secure/ViewProfile.jspa?name=koekemoe) for any testing that INS may want to do for this. It's available right now on jwst/master in github, but won't be in a public release until we start making releases for the upcoming B9.2 in 1-2 months. You can close the ticket when you want.

**Comment by stscijgbot-jp** [comment]
_Comment by [Anton Koekemoer](https://jira.stsci.edu/secure/ViewProfile.jspa?name=koekemoe) on [JIRA](https://jira.stsci.edu/browse/JP-2813?focusedCommentId=630258#comment-630258):_

Thanks. In order to test it, can you please provide the relevant python statements (necessary imports, name of the script, examples of how to run it, etc)?

**Comment by stscijgbot-jp** [comment]
_Comment by [Mihai Cara](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mcara) on [JIRA](https://jira.stsci.edu/browse/JP-2813?focusedCommentId=630270#comment-630270):_

The documentation for the command-line script is here: <https://github.com/spacetelescope/jwst/blob/master/jwst/scripts/adjust_wcs.py#L32-L48>  (which needs small correction which I will do soon) and the documentation for the Python function is here: <https://jwst-pipeline.readthedocs.io/en/latest/jwst/tweakreg/utils.html#jwst.tweakreg.utils.adjust_wcs>

For example, to apply a correction to the WCS of all '`{}cal{`}' files in a directory such as a DEC by 0.2 degrees, RA by 0.3 degrees, roll angle by 0.4, and scale by a factor of 1.005, you can run a bash command:

`% adjust_wcs *_cal.fits --suffix wcsadj -r 0.2 -d 0.3 -o 0.4 -s 1.005`

This will apply the same correction to all files '`{}_cal.fits{`}{*}'.{*} It will save modified models as '`{}_wcsadj.fits{`}' (replace suffix '`{}cal{`}' to '`{}wcsadj{`}'). If, however, you just want to update existing files (overwriting them), simply replace option '\-\-suffix wcsadj' with '\-\-update'.

**Comment by stscijgbot-jp** [comment]
_Comment by [Anton Koekemoer](https://jira.stsci.edu/secure/ViewProfile.jspa?name=koekemoe) on [JIRA](https://jira.stsci.edu/browse/JP-2813?focusedCommentId=630665#comment-630665):_

Thanks very much, that's very helpful. I'll give that a try and will let you know once I've had a chance to do various tests on it.

**Comment by stscijgbot-jp** [comment]
_Comment by [Mattia Libralato](https://jira.stsci.edu/secure/ViewProfile.jspa?name=libra) on [JIRA](https://jira.stsci.edu/browse/JP-2813?focusedCommentId=642875#comment-642875):_

[Mihai Cara](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mcara): I'm testing the command line script. How is this supposed to work? Let's say that I load the Gaia DR3 catalog on top on an image with ds9, are the delta values that the script needs just the difference between the Gaia position and the position of a star on the image that I can estimate by eye in ds9?

A first comment I have is that it might be useful to work in arcsec at least instead of deg. The delta values are usually very small, 10^-3 deg or less.

**Comment by stscijgbot-jp** [comment]
_Comment by [Mattia Libralato](https://jira.stsci.edu/secure/ViewProfile.jspa?name=libra) on [JIRA](https://jira.stsci.edu/browse/JP-2813?focusedCommentId=642882#comment-642882):_

OK, I think the answer to my question is yes. The first test was not good but the reason is that also the scale was off. After adjusting the scale, the result is much better. I do not have any data set to test a roll angle issue.

**Comment by stscijgbot-jp** [comment]
_Comment by [Tony Sohn](https://jira.stsci.edu/secure/ViewProfile.jspa?name=tsohn) on [JIRA](https://jira.stsci.edu/browse/JP-2813?focusedCommentId=643217#comment-643217):_

Reporting from the NIRISS imaging team.
 * For the testing of adjust_wcs script, I created a new conda environment named jwst-test and installed jwst version 1.10.2 through "`{}pip install jwst==1.10.2{`}". Image used for the test is located in: `/ifs/jwst/wit/niriss/distortion_data/1501/jw01501001001_24101_00001_nis_cal.fits`

 *  First test was done using small shifts. Due to pointing error, overlaying Gaia stars shows an overall slight shift against the observed positions. I eyeball-measured these shifts in RA, Dec, and used the followed command to shift the input image:

```java
`adjust_wcs jw01501001001_24101_00001_nis_cal.fits --suffix wcsadj1 -r -0.0001943 -d 0.0000201 -s 1.000````
 * The resulting image `jw01501001001_24101_00001_nis_wcsadj1.fits` (in the same dir above) was shifted in RA, Dec as expected, and shows a better match to Gaia stars. The attached `niriss_blink1.gif` animiation was created to compare how the WCS-shifted image now matches the Gaia catalog positions better.


 * Second test was done on the same image, but now with bigger shifts (1 arcmin in each direction) as well as rotation (10 deg) and scale (x1.2) applied. The command I used was:

```java
adjust_wcs jw01501001001_24101_00001_nis_cal.fits --suffix wcsadj2 -r 0.0167 -d 0.0167 -o 10 -s 1.2```
 * The resulting image `jw01501001001_24101_00001_nis_wcsadj2.fits` was shifted, rotated, and scaled as expected. The attached `niriss_blink2.gif` animation shows the before and after applying adjust_wcs images. To create this gif animation, the two images were "matched" in WCS in Saoimage DS9. 


 * *Summary: The adjust_wcs script ran successfully and created output images consistent with expectations.*

 * Further comments:

 * 
 ** I noticed that the input values for shift/rotation/scale does NOT accept scientific/engineering notations. For example, -0.0001943 works but -1.934e-4 does not. This may be inconvenient for users because in most cases, this script will be used for applying small adjustments – for example, a few arcseconds in shifts. If allowing scientific notation is difficult, an alternative could be to use arcsec units for the RA/Dec shifts.

 * 
 ** Not sure if this was intended, but using the --suffix option replaces the _cal.fits part instead of adding suffix at the end of filename. For example, the output is named `<img_base>_nis_wcsadj.fits` instead of `{}<img_base>_nis_cal_wcsadj.fits{`}.

**Comment by stscijgbot-jp** [comment]
_Comment by [Misty Cracraft](https://jira.stsci.edu/secure/ViewProfile.jspa?name=cracraft) on [JIRA](https://jira.stsci.edu/browse/JP-2813?focusedCommentId=645397#comment-645397):_

It sounds like NIRISS and MIRI testing succeeded and gave expected results. (Though both commented that maybe the units to use should be arcsec rather than degrees.) Is that everything needed for testing?

**Comment by stscijgbot-jp** [comment]
_Comment by [Anton Koekemoer](https://jira.stsci.edu/secure/ViewProfile.jspa?name=koekemoe) on [JIRA](https://jira.stsci.edu/browse/JP-2813?focusedCommentId=645404#comment-645404):_

Thanks for the update, I've tried this as well for NIRCam and it seems to work as advertised.

However, a couple of comments: (1) the fact that the input values can't accept scientific/engineering notation should be addressed, and (2) is there a way to call this from within python? (or if not, can it be made so please?) - this would solve the input unit issue easily since if we call it from within python, those values could presumably be passed using the astropy.units functionality.

**Comment by stscijgbot-jp** [comment]
_Comment by [Mihai Cara](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mcara) on [JIRA](https://jira.stsci.edu/browse/JP-2813?focusedCommentId=645489#comment-645489):_

The Python function exists and it is documented here: <https://jwst-pipeline.readthedocs.io/en/latest/jwst/tweakreg/utils.html#jwst.tweakreg.utils.adjust_wcs> However, it currently does not support units. Scientific notation is supported by the command line parser but it is the negative sign combined with 'e', I believe, that is causing an issue because it is interpreted as a new option instead of the argument of the previous option. There is a workaround to this and I will need to do some more work.

So, do not close this issue for now.

**Comment by stscijgbot-jp** [comment]
_Comment by [Mihai Cara](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mcara) on [JIRA](https://jira.stsci.edu/browse/JP-2813?focusedCommentId=650977#comment-650977):_

The issues of negative angles in engineering format and support for angle units have been addressed in  #7683
